internal/k8s/vcf_controller.go

```diff
@@ -0,0 +1,181 @@
+func (c *AviController) HandleVCF(informers K8sinformers, stopCh <-chan struct{}, ctrlCh chan struct{}) {
+	cs := c.informers.ClientSet
+	aviSecret, err := cs.CoreV1().Secrets(utils.GetAKONamespace()).Get(context.TODO(), lib.AviSecret, metav1.GetOptions{})
+	if err == nil {
+		ctrlIP := os.Getenv(utils.ENV_CTRL_IPADDRESS)
+		authToken := aviSecret.Data["authToken"]
+		username := aviSecret.Data["username"]
+		var transport *http.Transport
+		_, err = clients.NewAviClient(
+			ctrlIP, string(username), session.SetAuthToken(string(authToken)),
+			session.SetRefreshAuthTokenCallbackV2(utils.GetAuthtokenFromCache),
+			session.SetNoControllerStatusCheck, session.SetTransport(transport),
+			session.SetInsecure,
+		)
+		if err == nil {
+			utils.AviLog.Infof("Successfully connected to AVI controller using existing AKO secret")
+			return
+		} else {
+			utils.AviLog.Error("AVI controller initialization failed with err: %v", err)
+		}
+	}
+
+	utils.AviLog.Infof("Got error while fetching avi-secret: %v", err)
+
+	if !c.ValidBSData() {
+		utils.AviLog.Infof("Running in a VCF Cluster, but valid Bootstrap CR not found, waiting .. ")
+		startSyncCh := make(chan struct{})
+		c.AddNCPBootstrapEventHandler(informers, stopCh, startSyncCh)
+	L:
+		for {
+			select {
+			case <-startSyncCh:
+				break L
+			case <-ctrlCh:
+				return
+			}
+		}
+	}
+	utils.AviLog.Infof("NCP Bootstrap CR found, continuing AKO initialization")
+	c.UpdateAviSecret()
+}
+
+func (c *AviController) UpdateAviSecret() error {
+	secretName, ns, username := lib.GetBootstrapCRData()
+	cs := c.informers.KubeClientIntf.ClientSet
+	cs.CoreV1().Secrets(ns).Get(context.TODO(), secretName, metav1.GetOptions{})
+
+	var ncpSecret *corev1.Secret
+	var err error
+	ncpSecret, err = cs.CoreV1().Secrets(ns).Get(context.TODO(), secretName, metav1.GetOptions{})
+	if err != nil {
+		utils.AviLog.Warnf("Failed to get secret, got err: %v", err)
+		return err
+	}
+
+	var aviSecret corev1.Secret
+	aviSecret.ObjectMeta.Name = lib.AviSecret
+	aviSecret.Data = make(map[string][]byte)
+	aviSecret.Data["authtoken"] = []byte(ncpSecret.Data["authToken"])
+	aviSecret.Data["username"] = []byte(username)
+
+	_, err = cs.CoreV1().Secrets(utils.GetAKONamespace()).Get(context.TODO(), lib.AviSecret, metav1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		_, err = cs.CoreV1().Secrets(utils.GetAKONamespace()).Create(context.TODO(), &aviSecret, metav1.CreateOptions{})
+		if err != nil {
+			utils.AviLog.Warnf("Failed to create avi-secret, err: %v", err)
+			return err
+		}
+	}
+
+	_, err = cs.CoreV1().Secrets(utils.GetAKONamespace()).Update(context.TODO(), &aviSecret, metav1.UpdateOptions{})
+	if err != nil {
+		utils.AviLog.Warnf("Failed to update avi-secret, err: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+func (c *AviController) ValidBSData() bool {
+	utils.AviLog.Infof("Validating BS data")
+	cs := c.informers.ClientSet
+	secretName, ns, username := lib.GetBootstrapCRData()
```

Comment:
There is a possibility that the `GetBootstrapCRData` method will return empty values for any of these 3 return variables, should we handle them appropriately?

Reply:
Good catch. Updated.